### ISSUE TITLE
Exploring Performance Improvements with More Powerful Hardware

### ISSUE THREAD

**Issue by MichaelMartinez** [jamespilgrim/PiTrac#1]
I've been fascinated by your project and the potential it has to bring affordable golf launch monitoring to the masses. As I was reading through the documentation, I couldn't help but wonder if using more powerful hardware could improve the performance of PiTrac.

Specifically, I was thinking about using a SFF computer like a Dell, Lenovo or HP with more ram, a more powerful processor and possibility of a GPU. I have two of them running various workloads in my shop/home. I have an HP prodesk G5 with 16gb of ram and a 256GB SSD. It cost $99 with shipping and It controls a DIY CNC machine without breaking a sweat. The other SFF computer had similar specs, but I installed an Nvidia 1050ti bringing the total cost to $150. My son games on this computer and doesn't complain about lagging.   

A question that keeps cropping up in my mind is; Would running the cameras directly to a computer with a GPU potentially unlock CUDA or ROCm acceleration, thus significantly boosting processing speeds?

I understand that one of the goals of PiTrac is to keep costs low, but I'm curious to know if you've explored these options or considered the potential benefits they could bring. Would using more powerful hardware allow for:

- Faster processing of video feeds?
- Improved accuracy in tracking golf ball launch speed, angles, and spin?
- Enhanced overall system responsiveness?

I'd love to hear your thoughts on this and any experiments you may have conducted with different hardware configurations.

Considerations:

- I've read through the project logs and understand that the current implementation uses Raspberry Pi computers and cameras.
- I'm aware of the cost constraints and the goal of keeping PiTrac affordable, but I believe it's worth exploring potential performance improvements.

I look forward to hearing your thoughts on this!

**Comment by jamespilgrim** [comment]
Interesting - There are a number of areas where I think a stronger processer would help.  Will have to give this some more consideration, but here's some initial thoughts.

First, given that many people will use a golf simulator package like TruGolf or GSPro, it's likely that they will already have a more capable machine with a decent GPU anyway.  So, the incremental cost might not be that much.

Your thought about latency reduction is probably the biggest potential improvement.  The current spin analysis (and much of the circle detection before that) is pretty expensive.  A more-pipelined resource like a GPU could probably greatly improve the time between a shot being taken and the results coming up.   Accuracy might be improved, because we could use a larger number of candidate ball-spin images, resulting in a smaller angular increment between those images and hopefully more precision.

We haven't really done any experiments with other configurations yet.  So far, the whole focus has just been on getting something that works (mostly!) out into the world.  I'm really looking forward to being able to sit back a bit and just tinker with things, including different hardware.  And there's no reason there can't be variations of PiTracs from budget to higher-end in time.

We mostly just wanted to first prove that this could be done, and now we can have some fun seeing where it goes!

**Comment by jeshernandez** [comment]
Interesting. I was actually thinking the opposite and reducing to one raspberry pi 5 with 8GB Ram and an added NVidia card for more GPU capability (https://www.youtube.com/watch?v=AyR7iCS7gNI) to hopefully make it a smaller footprint and keep the cost down. However, I suppose that's the beauty of an open-source project. Once you get it running, there could be options for everyone. In my case, I simply want to be able to play Golf in my garage with satisfactory latency (under 3 seconds)? I can't wait to get something working and try different options!

**Comment by MichaelMartinez** [comment]
I appreciate your time and insight into the potential improvements that more powerful hardware could bring. Your points about latency and accuracy are dead on in my estimation, and I am excited about the possibility of seeing these improvements come to fruition.

As I understand, the bulk of the project code is written in C++, which I'm not proficient in; however, I am more than willing to learn and contribute or add value via Python, wherever possible. Additionally, the project log on Hackaday describes current issues with keys, vendor IDs, and other environment secrets that are sensitive. I completely understand the concern and the need to protect this information.

Given these issues, would it be possible to release a vendor-agnostic version of PiTrac that simply measures the golf ball's parameters? Such a version would allow experimenters like myself to at least get a system built and start contributing in one form or another. As long as your plans aren't moving towards commercialization of the project, I believe there's merit in the idea of "Don't let perfection be the enemy of good enough!" This would also give the community a chance to tinker, test, and possibly even offer valuable feedback and improvements.

I'm really excited about the direction PiTrac is heading and look forward to any updates or further discussions on exploring these options.

**Comment by jamespilgrim** [comment]
That's a good point, Michael.  I'm working today to see if I can get a
clean meson/ninja setup that will allow us to easily build without any of
the vendor-specific code that's causing the problem.

As an aside, the perfection-as-the-enemy is well taken.  Been struggling
with that for weeks.  We're mostly just worried about putting something out
there that folks won't be successful with and then sort of ruining our one
chance.  But if we never put anything out, that's probably just as bad. :/

More soon.

**Comment by MichaelMartinez** [comment]
> We're mostly just worried about putting something out there that folks won't be successful with and then sort of ruining our one chance. But if we never put anything out, that's probably just as bad. :/ More soon.
> 

I hope that this issue isn't being taken as pressure to do "something." My only intention was to provide feedback, hopefully constructive feedback. Please don't rush or feel obliged to do anything on my behalf. You have already inspired a plethora of ideas with what you have posted. So take your time and release when you feel the time is right... the universe will let you know with good momentum.

**Comment by jamespilgrim** [comment]
No worries.  I wanted to get something out soon in any case.  Please see
this announcement
<https://hackaday.io/project/195042-pitrac-the-diy-golf-launch-monitor/log/237457-first-release-of-pitrac-source-code>!
:)

**Comment by benhalpern** [comment]
+1 on vendor-agnostic interface to pipe out all the data and allow a connective program (written in Python or whatever, as @MichaelMartinez describes) to stitch together additional ways to use the data produced.

It probably makes sense to support a few direct connections in the core code (if I understand how this works correctly), to "major" vendors, but to mostly try and keep the core minimal in what it interfaces with and mostly restrict what the core project is responsible for to tasks which require high performance — allowing higher-level programs to handle other tasks.

> I was actually thinking the opposite and reducing to one raspberry pi 5 with 8GB Ram and an added AMD card for more GPU capability (https://www.youtube.com/watch?v=AyR7iCS7gNI) to hopefully make it a smaller footprint and keep the cost down. However

I also agree with this — and generally I'd think there should be an effort to shrink the core as much as possible while still being functional out of the box, but strive for a modular/pluggable architecture for performance increases — and even strive for a case which could be accepting of additional parts which plug in (though this is an area I'm pretty clueless on as far as details go 😄).

One final thought from the announcement:

> The code drop is not quite ready for prime time, but we didn't want another weekend to go by without giving folks a chance to compile and see the source code.

Huge +1 to this. It's never going to feel ready. Maybe the project could come up with some official vernacular to express the state of different parts of the project and what is most settled vs likely to change?

**Comment by jamespilgrim** [comment]
That reminds me - we actually DO already have a fully agnostic interface! . It’s built on ActiveMQ and so can interface to any JMS client,  CORBA etc. over a number of wire protocols. Just need to document it. The idea was to enable third-party add-one, like team oriented networks of PiTracs, club-oriented aops and such.  I’ll see if I can create some early documentation this week. The PiTrac web GUI is one such application (Java/Tomcat) and we have already pushed that source code out. On Dec 30, 2024, at 8:22 AM, Ben Halpern ***@***.***> wrote:﻿
+1 on vendor-agnostic interface to pipe out all the data and allow a connective program (written in Python or whatever, as @MichaelMartinez describes) to stitch together additional ways to use the data produced.
It probably makes sense to support a few direct connections in the core code (if I understand how this works correctly), to "major" vendors, but to mostly try and keep the core minimal in what it interfaces with and mostly restrict what the core project is responsible for to tasks which require high performance — allowing higher-level programs to handle other tasks.

I was actually thinking the opposite and reducing to one raspberry pi 5 with 8GB Ram and an added AMD card for more GPU capability (https://www.youtube.com/watch?v=AyR7iCS7gNI) to hopefully make it a smaller footprint and keep the cost down. However

I also agree with this — and generally I'd think there should be an effort to shrink the core as much as possible while still being functional out of the box, but strive for a modular/pluggable architecture for performance increases — and even strive for a case which could be accepting of additional parts which plug in (though this is an area I'm pretty clueless on as far as details go 😄).
One final thought from the announcement:

The code drop is not quite ready for prime time, but we didn't want another weekend to go by without giving folks a chance to compile and see the source code.

Huge +1 to this. It's never going to feel ready. Maybe the project could come up with some official vernacular to express the state of different parts of the project and what is most settled vs likely to change?

**Comment by benhalpern** [comment]
Yeah I think the existing agnostic piping seems like the way to go as a reliable dev platform as needed and just needs to be documented. As long as it's available and functional, people can build their own abstractions to tailor to necessary use cases.
